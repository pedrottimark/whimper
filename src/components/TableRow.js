/* @flow */

import React, {Component} from 'react';
import invariant from 'invariant';

import type {
  DeleteRecord,
  Fields,
  Record,
  Updating,
} from '../types';

type Props = {|
  deleteRecord: DeleteRecord,
  fields: Fields,
  record: Record,
  updating: Updating,
|};

class TableRow extends Component {
  props: Props;
  _input: HTMLInputElement;

  _refInput = (input: HTMLInputElement) => {
    this._input = input;
  }

  _onSubmit = (e: Event) => {
    e.preventDefault();
    invariant(this.props.updating, 'updating is defined when TableRow renders a form');
    this.props.updating.update(this._input.value);
  }

  render() {
    const {deleteRecord, fields, record, updating} = this.props;
    const recordId = record.id;
    const symbol = '\u274c'; // CROSS MARK
    return (
      <tr>
        <td key="" title="delete row" onClick={deleteRecord.bind(null, recordId)}><button>{symbol}</button></td>
        {
          fields.map(({key, type}) => {
            const value = record[key];
            return (
              <td key={key} data-record-id={recordId} data-field-key={key}>
                {
                  updating && updating.recordId === recordId && updating.fieldKey === key
                    ? <div>
                        <span>{value}</span>
                        <form onSubmit={this._onSubmit}>
                          <input ref={this._refInput} type={type} defaultValue={value} autoFocus={true} />
                        </form>
                      </div>
                    : value
                }
              </td>
            );
          })
        }
      </tr>
    );
  }
}

export default TableRow;
