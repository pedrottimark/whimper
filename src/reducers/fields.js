/* @flow */

import type {
  Action,
  Field,
  Fields,
  FieldValue,
} from '../types';

// production app might get fields from external source :)
export const fieldsReceived: Fields = [
  {
    key: 'when',
    label: 'when',
    type: 'number',
    defaultValue: new Date().getFullYear(),
  },
  {
    key: 'what',
    label: 'what',
    type: 'text',
    defaultValue: '',
  },
  {
    key: 'weep',
    label: 'weep',
    type: 'text',
    defaultValue: '',
  },
];

// Given a field and text input value,
// return a data value to update the property in a record.
export const fieldValue = ({type}: Field, text: string): FieldValue => type === 'number'
  ? parseInt(text, 10)
  : text;

export const fieldsInitial: Fields = [];

export default function (fields: Fields = fieldsInitial, action: Action): Fields {
  switch (action.type) {

  case 'RECEIVE_DATA':
    return action.fields || fields;

  default:
    return fields;

  }
}
