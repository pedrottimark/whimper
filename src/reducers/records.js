/* @flow */

import type {
  Action,
  DeleteRecordAction,
  UpdateFieldAction,
  Fields,
  Record,
  RecordId,
  Records,
} from '../types';

/*
// Convert internal collection from and to external JSON.
export const recordsFromJSON = (array: Record[]): Records => array;
export const recordsToJSON = (records: Records): Record[] => records;
*/

// Initial state of records before they have been received from external source.
export const recordsInitial: Records = [];

// A record object has an implicit and temporary id: its index in the array.
export default function (records: Records = recordsInitial, action: Action): Records {
  switch (action.type) {

  case 'CREATE_RECORD':
    // Create new record as first element of the collection.
    return [action.record].concat(records);

  case 'RECEIVE_DATA':
    // Replace initial records, except when whimper starts for the first time.
    return action.records || records;

  case 'UPDATE_FIELD': {
    // Destructuring and type cast to avoid warnings in Flow 0.35.0
    const {recordId, fieldKey, value} = (action: UpdateFieldAction);
    return records.map(record => record.id === recordId
      ? Object.assign({}, record, {[fieldKey]: value})
      : record
    );
  }

  case 'DELETE_RECORD': {
    // Destructuring and type cast to avoid warnings in Flow 0.35.0
    const {recordId} = (action: DeleteRecordAction);
    return records.filter(record => record.id !== recordId);
  }

  default:
    return records;

  }
}

// A very naive algorithm, instead of hashing ;)
const unusedId = (records: Records): RecordId =>
  records.reduce((id, record) => id <= record.id ? record.id + 1 : id, 0);

// Given fields, return a record with default values.
// For example, to create a new record.
export const recordDefault = (fields: Fields, records: Records): Record =>
  fields.reduce((record, {key, defaultValue}) => {
    record[key] = defaultValue;
    return record;
  }, {
    id: unusedId(records)
  });
