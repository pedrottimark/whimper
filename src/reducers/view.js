import type {
  Action,
  FieldValue,
  Fields,
  FilteringCriterion,
  Record,
  View,
} from '../types';

const filteringInitial = '';
const sortingInitial = [];
export const viewInitial: View = {
  filtering: filteringInitial,
  sorting: sortingInitial,
};

export default function (view: View = viewInitial, action: Action): View {
  switch (action.type) {

    case 'SEARCH_RECORDS':
      return Object.assign({}, view, {
        filtering: action.filtering
      });

    case 'SORT_RECORDS':
      // Add current sorting criteria as first element of array.
      // Toggle descending if previous first element has same fieldKey.
      // The array can have at most one element for any fieldKey.
      return Object.assign({}, view, {
        sorting: action.fieldKey
          ? [{
              fieldKey: action.fieldKey,
              descending: view.sorting.length !== 0 &&
                view.sorting[0].fieldKey === action.fieldKey &&
                !view.sorting[0].descending,
            }].concat(view.sorting.filter(({fieldKey}) => fieldKey !== action.fieldKey))
          : sortingInitial
      });

    default:
      return view;
  }
}

// Return sorting comparision of field values.
const compareFieldValues = (fieldValueA: FieldValue, fieldValueB: FieldValue, descending: boolean): number => {
  const comparison = typeof fieldValueA === 'number' && typeof fieldValueB === 'number'
    ? fieldValueA - fieldValueB
    : String(fieldValueA).localeCompare(String(fieldValueB));

  return descending ? -comparison : comparison;
};

// Return sorting comparision of records for zero or more fields.
// Assume that the sorting is stable (for example, OrderedMap immutable collection).
const compareRecords = (recordA: Record, recordB: Record, sorting: SortingCriterion[]) : number =>
  sorting.reduce((comparison, {fieldKey, descending}) =>
    comparison || // compare field values only if all preceding comparisons are zero
    compareFieldValues(recordA[fieldKey], recordB[fieldKey], descending),
  0);

// Return whether the record has at least one field which includes the filtering string.
const filterRecord = (record: Record, fields: Fields, filtering: FilteringCriterion): boolean =>
  filtering.length === 0 || // not filtering
  fields.some(({key: fieldKey}) => record[fieldKey].toString().toLowerCase().includes(filtering));

// Return records to display in a view, given the sorting and filtering critera.
// A record id is its index in the currently stored list of records.
export const recordsInView = (records: Records, fields: Fields, {filtering, sorting}: View): Records =>
  records
    .filter((record) => filterRecord(record, fields, filtering))
    .sort((recordA, recordB) => compareRecords(recordA, recordB, sorting));
