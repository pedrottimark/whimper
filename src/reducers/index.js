import { combineReducers } from 'redux';

// child reducers
import fields from './fields';
import records from './records';
import view from './view';

// root reducer
export default combineReducers({
  fields,
  records,
  view,
});
