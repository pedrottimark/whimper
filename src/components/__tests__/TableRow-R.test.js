/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import invariant from 'invariant';

import {fieldsReceived} from '../../reducers/fields';
import {recordA as record} from '../../testing/records-data';

import TableRow from '../TableRow';

describe('TableRow', () => {
  it('renders a record', () => {
    const updating = null;

    expect(renderer.create(
      <TableRow
        deleteRecord={jest.fn()}
        fields={fieldsReceived}
        record={record}
        updating={updating}
      />
    ).toJSON()).toMatchSnapshot();
  });

  it('renders a record updating a number field', () => {
    const field = fieldsReceived[0];
    invariant(field.type === 'number', 'testing a number field');
    const updating = {
      recordId: record.id,
      fieldKey: field.key,
      update: jest.fn(),
    };

    expect(renderer.create(
      <TableRow
        deleteRecord={jest.fn()}
        fields={fieldsReceived}
        record={record}
        updating={updating}
      />
    ).toJSON()).toMatchSnapshot();
  });

  it('renders a record updating a text field', () => {
    const field = fieldsReceived[1];
    invariant(field.type === 'text', 'testing a text field');
    const updating = {
      recordId: record.id,
      fieldKey: field.key,
      update: jest.fn(),
    };

    expect(renderer.create(
      <TableRow
        deleteRecord={jest.fn()}
        fields={fieldsReceived}
        record={record}
        updating={updating}
      />
    ).toJSON()).toMatchSnapshot();
  });
});
