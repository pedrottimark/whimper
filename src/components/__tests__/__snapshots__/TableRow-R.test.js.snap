// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow renders a record 1`] = `
<tr>
  <td
    onClick={[Function]}
    title="delete row"
  >
    <button>
      ❌
    </button>
  </td>
  <td
    data-field-key="when"
    data-record-id={1}
  >
    2017
  </td>
  <td
    data-field-key="what"
    data-record-id={1}
  >
    ECMAScript 2017
  </td>
  <td
    data-field-key="weep"
    data-record-id={1}
  >
    async or swim, not much longer to await
  </td>
</tr>
`;

exports[`TableRow renders a record updating a number field 1`] = `
<tr>
  <td
    onClick={[Function]}
    title="delete row"
  >
    <button>
      ❌
    </button>
  </td>
  <td
    data-field-key="when"
    data-record-id={1}
  >
    <div>
      <span>
        2017
      </span>
      <form
        onSubmit={[Function]}
      >
        <input
          autoFocus={true}
          defaultValue={2017}
          type="number"
        />
      </form>
    </div>
  </td>
  <td
    data-field-key="what"
    data-record-id={1}
  >
    ECMAScript 2017
  </td>
  <td
    data-field-key="weep"
    data-record-id={1}
  >
    async or swim, not much longer to await
  </td>
</tr>
`;

exports[`TableRow renders a record updating a text field 1`] = `
<tr>
  <td
    onClick={[Function]}
    title="delete row"
  >
    <button>
      ❌
    </button>
  </td>
  <td
    data-field-key="when"
    data-record-id={1}
  >
    2017
  </td>
  <td
    data-field-key="what"
    data-record-id={1}
  >
    <div>
      <span>
        ECMAScript 2017
      </span>
      <form
        onSubmit={[Function]}
      >
        <input
          autoFocus={true}
          defaultValue="ECMAScript 2017"
          type="text"
        />
      </form>
    </div>
  </td>
  <td
    data-field-key="weep"
    data-record-id={1}
  >
    async or swim, not much longer to await
  </td>
</tr>
`;
